src/Shared/CoreCLRAssemblyLoader.cs

```diff
@@ -57,113 +28,23 @@ public Assembly LoadFromPath(string fullPath)
 
             lock (_guard)
             {
-                if (!_resolvingHandlerHookedUp)
-                {
-                    AssemblyLoadContext.Default.Resolving += TryResolveAssembly;
-                    _resolvingHandlerHookedUp = true;
-                }
-
                 Assembly assembly;
                 if (_pathsToAssemblies.TryGetValue(fullPath, out assembly))
                 {
                     return assembly;
                 }
 
-                return LoadAndCache(AssemblyLoadContext.Default, fullPath);
-            }
-        }
+                var contextForAssemblyPath = new MSBuildLoadContext(fullPath);
 
-        private Assembly TryGetWellKnownAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
-        {
-            if (!_wellKnownAssemblyNames.Contains(assemblyName.Name))
-            {
-                return null;
-            }
-
-            // Ensure we are attempting to load a matching version
-            // of the Microsoft.Build.* assembly.
-            assemblyName.Version = _currentAssemblyVersion;
-
-            var searchPaths = new[] { Assembly.GetExecutingAssembly().Location };
-            return TryResolveAssemblyFromPaths(context, assemblyName, searchPaths);
-        }
-
-        private Assembly TryResolveAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
-        {
-            lock (_guard)
-            {
-                Assembly assembly = TryGetWellKnownAssembly(context, assemblyName);
+                assembly = contextForAssemblyPath.LoadFromAssemblyPath(fullPath);
 
                 if (assembly != null)
```

Comment:
Shouldn't `null` loads be recorded as well? If not failed loads will just be attempted over and over again correct?

Reply:
Well - runtime caches the results of bind operations (including failures), so overall this should not be called too often really. That said if the cache is already in place, it would definitely help to use if for all assemblies, but I would not expect the perf gain to be substantial.

Reply:
Looking around, it looks like there are static fields that keep a `CoreCLRAssemblyLoader` around for the lifetime of the process, which for a worker node could be long. If you:

1. build
1. it fails to load a task because you didn't restore so the file doesn't exist (or something)
1. restore
1. build

We need to try to load again in step 4 even if it failed in step 2.

(Did I make the change and get 80% though typing up a commit message justifying it before realizing this? Yes I did.)

Reply:
Is there a conflict between above needing to work and the comment just above that says runtime caches failures?